docs/apis/subsystems/roles.md

```diff
@@ -0,0 +1,141 @@
+
+See [Roles and modules#Context](https://docs.moodle.org/dev/Roles_and_modules#Context) for more information.
+
+### Capabilities
+
+Capabilities can have the following permissions:
+
+1. `CAP_INHERIT`
+2. `CAP_ALLOW`
+3. `CAP_PREVENT`
+4. `CAP_PROHIBIT`
+
+If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.
+
+### Capability conflicts
+
+Since the capabilities in each role could be different, there could be conflict in capabilities.
+
+If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".
+
+## Programming Interface
+
+Although the Roles system may look complicated at first glance, implementing it in Moodle code is fairly simple.
+
+- Examples of functions handling the **Roles** can be found at [`accesslib.php`](https://github.com/moodle/moodle/blob/master/lib/accesslib.php)
+
+- Moodle comes with a list of predefined roles for example Student, Teacher, Course Manager etc. A new role can be created based on
+    those predefined roles using **role archetype:**
+
+```php
+/**
+ * Returns array of all role archetypes.
+ *
+ * @return array
+ */
+function get_role_archetypes() {
+    return array(
+        'manager'        => 'manager',
+        'coursecreator'  => 'coursecreator',
+        'editingteacher' => 'editingteacher',
+        'teacher'        => 'teacher',
+        'student'        => 'student',
+        'guest'          => 'guest',
+        'user'           => 'user',
+        'frontpage'      => 'frontpage'
+    );
+}
+```
+
+- On each page you need to find the context the user is working in, using the `get_context_instance()` function.  For example, in the forum module:
+
+```php
+$context = get_context_instance(CONTEXT_MODULE, $cm->id);
+```
+
+or at the course level:
+
+```php
+$context = get_context_instance(CONTEXT_COURSE, $id);
+```
+
+- Then, whenever you want to check which role has a given capability do something like this:
+
+```php
+list($roleids) = get_roles_with_cap_in_context($context, 'moodle/course:manageactivities');
+```
+
+- If you want to get a list of roles that a give user has you can call
+
+```php
+$ras = get_user_roles($context, $user->id);
+```
+
+- You can assign a role to a user by calling something like
+
+```php
+role_assign($roleid, $userid, $context->id);
+```
+
+## See also
+
+- [Hardening new Roles system](https://docs.moodle.org/dev/Hardening_new_Roles_system)
```

Comment:
Let's remove the word 'new' - it's been over 10 years
```suggestion
- [Hardening the Roles system](https://docs.moodle.org/dev/Hardening_new_Roles_system)
```

Ideally we should merge that document into this one.